Importing command from text (json file)

BartlomiejJaruga opened the pull request:
feat:
- added CommandLoader with static method for loading List of DriverCommands from file
- added Test for CommandLoader in ComplexCommandTest.java
- added example file for test (fileWithTestCommands.json)
- added input and button for loading commands from file in GUI in CommandManagerWindow
- added jackson dependency needed for mapping data from json to java class

BartlomiejJaruga committed:
feat:
- added CommandLoader with static method for loading List of DriverCommands from file
- added Test for CommandLoader in ComplexCommandTest.java
- added example file for test (fileWithTestCommands.json)
- added input and button for loading commands from file in GUI in CommandManagerWindow
- added jackson dependency needed for mapping data from json to java class

kowallus reviewed (changes requested):
General lack of S.O.L.I.D.

BartlomiejJaruga committed:
Merge branch 'iis-powp-2025:master' into master

BartlomiejJaruga committed:
refactor:
- changed CommandLoader to static utility class CommandParser which now has only one function to parse List of CommandEntry to List of DriverCommand
- removed unnecessary dependantions in CommandParser - no more mapper from file there
- JsonCommandEntry now implements CommandEntry
- refactored ComplexCommandTest to work with current version of CommandParser

feature:
- added CommandEntryMapper Interface used to introduce loading commands from file with new extension (.json)
- CommandManagerWindow has new dependency injection - CommandEntryMapper
- CommandManagerWindow method loadCommandsFromFile now uses CommandEntryMapper injected through constructor and utility class CommandParser instead of old multi-dependency class CommandLoader

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/entries/JsonCommandEntry.java:
```diff
@@ -0,0 +1,19 @@
+package edu.kis.powp.jobs2d.command.entries;
+
+public class JsonCommandEntry implements CommandEntry {
```
What is so "Json" about this class? 
How would XMLCommandEntry look like?
I think I don't understand the idea of CommandEntry hierarchy.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/mapper/JsonCommandEntryMapper.java:
```diff
@@ -0,0 +1,26 @@
+    public List<CommandEntry> mapToCommandEntries(String filename) {
+        try{
+            File file = new File(filename);
+            List<JsonCommandEntry> jsonCommands = mapper.readValue(file, new TypeReference<List<JsonCommandEntry>>() {});
```
Not fixed:
"Reading directly from file, disables OCP (mapper can also parse strings). It also brakes SRP of this class (reading from a file, and parsing from json)"

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/mapper/CommandEntryMapper.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.List;
+
+public interface CommandEntryMapper {
+    ObjectMapper mapper = new ObjectMapper();
```
interface depends on com.fasterxml.jackson.databind.ObjectMapper, why?!

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/mapper/CommandEntryMapper.java:
```diff
@@ -0,0 +1,13 @@
+public interface CommandEntryMapper {
+    ObjectMapper mapper = new ObjectMapper();
+
+    List<CommandEntry> mapToCommandEntries(String filename);
```
Why "filename" instead of: List<CommandEntry> mapToCommandEntries(String text);

kowallus reviewed (changes requested):
more SRP and OCP please

BartlomiejJaruga committed:
refactor:
- Mapper classes replaced with ManualJsonParser utility class
- Removed unnecessary JsonCommandEntry class which implemented CommandEntry interface
- Changed CommandEntry from interface to class
- renamed GUI components to fit new way of importing commands (from file to text string of json format)
- changed the way of getting commands imported - now we pass String input in format of json instead of giving file path
- removed completely importing commands from file
- removed unused jackson dependency

feature:
- added ManualJsonParser for manual parsing text to CommandEntry object

BartlomiejJaruga commented:
### Summary description:
We changed the approach of loading commands - now we parse commands from String rather than loading from file.
We removed jackson dependecies and now unusued mappers.
These changes lead to improving OSP and SRP.

### Refactor:
- Replaced Mapper classes with a new utility class ManualJsonParser for manual JSON parsing.
- Removed the unnecessary JsonCommandEntry class that implemented CommandEntry.
- Changed CommandEntry from an interface to a concrete class for simplification.
- Updated the command import mechanism: now accepts raw JSON string input instead of a file path.
- Completely removed file-based command import functionality.
- Renamed GUI components to reflect the new input method (from file-based to JSON text-based).
- Removed unused jackson dependency (no longer needed for JSON parsing).

### Feature:
- Introduced ManualJsonParser to manually parse JSON text into CommandEntry objects.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/gui/CommandManagerWindow.java:
```diff
@@ -324,4 +333,20 @@ private void editCurrentCommand() {
+        String commandsText = commandsInputTextField.getText();
+
+        try {
+            List<CommandEntry> commands = ManualJsonParser.parseCommands(commandsText);
```
No OCP and DIP again. CommandManagerWindow depends on json format and there is no room for extension without modification.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/ManualJsonParser.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.List;
+
+public class ManualJsonParser {
+    public static List<CommandEntry> parseCommands(String json) {
```
No room for Dependency inversion and OCP here. Only json strings support without modifications. 
Hint: DIP requires introducing interfaces.

kowallus reviewed (changes requested):
One of the goals of the course is to understand SOLID. IMO: you still fail to introduce OCP and DIP.

BartlomiejJaruga committed:
refactor:
- improved OCP and DIP

BartlomiejJaruga committed:
feature:
- introduced Strategy design pattern
- added CommandParsingContext class that is injected by constructor to CommandManagerWindow and used to automatically detect best strategy for input text
- added ParsingStrategy Interface in order to use Strategy design pattern
- added JsonParsingStrategy class implementing ParsingStrategy to test if the strategy pattern works with one strategy

BartlomiejJaruga committed:
fix:
- fixed CommandParsingContext auto-detect feature to reset setStategy to null after parsing commands

feature:
- added CsvParsingStrategy class implementing ParsingStrategy to test if the strategy pattern works with more strategies
- added combobox for selecting different Parsing options for importing commands

delete:
- removed unused classes (ManualJsonParser and ManualParser)

BartlomiejJaruga commented:
## This PR introduces a refactor of the command parsing mechanism using the Strategy design pattern. 

A new CommandParsingContext class was added and is injected into CommandManagerWindow, allowing automatic detection of the appropriate parsing strategy based on input. 
To support this design, a ParsingStrategy interface was introduced, along with two concrete implementations: JsonParsingStrategy and CsvParsingStrategy. 

A combo box was also added to the UI to enable manual selection of the desired parsing strategy (auto-detect, JSON, CSV). 
New strategies are automatically added to combo box without need of changing any code in CommandManagerWindow.

As part of the code cleanup, obsolete classes ManualJsonParser and ManualParser were removed.

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -45,10 +38,8 @@ public class TestJobs2dApp {
-                DriverFeature.getDriverManager());
+        SelectTestFigureOptionListener selectTestFigureOptionListener = new SelectTestFigureOptionListener(DriverFeature.getDriverManager());
+        SelectTestFigure2OptionListener selectTestFigure2OptionListener = new SelectTestFigure2OptionListener(DriverFeature.getDriverManager());
 
         application.addTest("Figure Joe 1", selectTestFigureOptionListener);
         application.addTest("Figure Joe 2", selectTestFigure2OptionListener);
```
It would be best if changes to the code unrelated to the PR would be in a separate PR. It is difficult to focus on important stuff.

BartlomiejJaruga replied:
We are very sorry, but these changes were caused by Intellij code auto-formatting. We haven't noticed it.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandParsingContext.java:
```diff
@@ -0,0 +1,69 @@
+        this.availableStrategies = new ArrayList<>();
+        // Register default strategies
+        registerStrategy(new JsonParsingStrategy());
+        registerStrategy(new CsvParsingStrategy());
```
If the code follows DIP this should never occur here. Dependency to Json and Csv routines are now propagated to CommandManagerWindow.
Hint: configure CommandParsingContext in TestJobs2dApp.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandParsingContext.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class CommandParsingContext {
```
Why Context?

BartlomiejJaruga replied:
We took the term `Context` from this website about strategy design pattern: [website](https://refactoring.guru/design-patterns/strategy).

The term `Context` reflects the role of the class in managing the parsing environment — storing available strategies and selecting the appropriate one dynamically based on input. It acts as a context for the Strategy pattern, maintaining the current state and configuration for parsing commands.

We have changed the name to `CommandParsingStrategiesManager` as suggested and to reflect the logic of storing all strategies and 'managing them'.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/command/CommandParsingContext.java:
```diff
@@ -0,0 +1,69 @@
+
+import edu.kis.powp.jobs2d.command.entries.CommandEntry;
+import edu.kis.powp.jobs2d.command.strategy.CsvParsingStrategy;
+import edu.kis.powp.jobs2d.command.strategy.JsonParsingStrategy;
```
Unnecessary and unwanted dependencies.

kowallus reviewed (changes requested):
Almost there. Fix one issue related with DIP and add files with example commands in json and csv in the folder src/test/resources (maven convention)

BartlomiejJaruga committed:
refactor:
- changed CommandParsingContext class name to CommandParsingStrategiesManager
- improved DIP by creating CommandParsingStrategiesManager constructor which takes List of ParsingStrategy with all strategies (TestJobs2dApp creates ready to go CommandParsingStrategiesManager)
- removed unnecessary dependencies (importing CsvParsingStrategy and JsonParsingStrategy in CommandParsingStrategiesManager)

BartlomiejJaruga committed:
fix:
- fixed tests for JsonCommandImport and CsvCommandImport
- fixed CommandParsingStrategiesManager constructor

feature:
- added .csv and .json example commands files in src/test/resources
